src/features/semanticTokensProvider.ts

```diff
@@ -0,0 +1,315 @@
+    XmlLiteralText,
+    RegexComment,
+    RegexCharacterClass,
+    RegexAnchor,
+    RegexQuantifier,
+    RegexGrouping,
+    RegexAlternation,
+    RegexText,
+    RegexSelfEscapedCharacter,
+    RegexOtherEscape,
+}
+
+enum SemanticHighlightModifier {
+    Static
+}
+
+export default class SemanticTokensProvider extends AbstractProvider implements vscode.DocumentSemanticTokensProvider, vscode.DocumentRangeSemanticTokensProvider {
+
+    getLegend(): vscode.SemanticTokensLegend {
+        return new vscode.SemanticTokensLegend(tokenTypes, tokenModifiers);
+    }
+
+    async provideDocumentSemanticTokens(document: vscode.TextDocument, token: vscode.CancellationToken): Promise<vscode.SemanticTokens | null> {
+
+        return this._provideSemanticTokens(document, null, token);
+    }
+
+    async provideDocumentRangeSemanticTokens(document: vscode.TextDocument, range: vscode.Range, token: vscode.CancellationToken): Promise<vscode.SemanticTokens | null> {
+
+        const v2Range: protocol.V2.Range = {
+            Start: {
+                Line: range.start.line,
+                Column: range.start.character
+            },
+            End: {
+                Line: range.end.line,
+                Column: range.end.character
+            }
+        };
+        return this._provideSemanticTokens(document, v2Range, token);
+    }
+
+    async _provideSemanticTokens(document: vscode.TextDocument, range: protocol.V2.Range, token: vscode.CancellationToken): Promise<vscode.SemanticTokens | null> {
+        let req = createRequest<protocol.V2.SemanticHighlightRequest>(document, new vscode.Position(0, 0));
+        req.Range = range;
+
+        let response = await serverUtils.getSemanticHighlights(this._server, req);
+
+        const semanticSpans = response.Spans;
+
+        const builder = new vscode.SemanticTokensBuilder();
+        for (let span of semanticSpans) {
+            let tokenType = tokenTypeMap[span.Type];
+            if (tokenType === undefined) {
+                continue;
+            }
+
+            let tokenModifiers = span.Modifiers.reduce((modifiers, modifier) => modifiers + tokenModifierMap[modifier], 0);
+
+            // We could add a separate classification for constants but they are
+            // supported as a readonly variable. Until we start getting more complete
+            // modifiers from the highlight service we can add the readonly modifier here.
+            if (span.Type === SemanticHighlightClassification.ConstantName) {
+                tokenModifiers += 2 ** DefaultTokenModifier.readonly;
+            }
+
+            var spanRange = toRange2(span);
+            for (let line = spanRange.start.line; line <= spanRange.end.line; line++) {
+                const startCharacter = (line === spanRange.start.line ? spanRange.start.character : 0);
+                const endCharacter = (line === spanRange.end.line ? spanRange.end.character : document.lineAt(line).text.length);
```

Comment:
Hmmm, you're making the assumption that document.lineAt(line - 1).text is equivalent to what the language server saw right? This is after several await's, at this point it's super possible the document representation has changed by now. Typically VSCode will map-forward whatever results you provide to the latest document they know about because they're aware of what doc version they originally requested

Reply:
This is all based off the assumption that vscode.TextDocument is mutable. Which i'm pretttty sure it is 😄 . I don't believe they run on a snaphsot system

Reply:
Thanks for pointing this out. I added back a version check that was present in the TypeScript Extensions implementation.